File: python/taflow/indicators/garman_klass.py
"""Native-backed Garman-Klass volatility adapter."""

from typing import Any

import numpy as np

from .._native import GarmanKlassOperator as _Native
from .._adapter_protocol import adapter_length
from .._series import as_float64_series


class GarmanKlass:
    """Compute rolling Garman-Klass OHLC volatility.

    ``_open``, ``high``, ``low``, and ``close`` are required aligned
    chronological histories; four empty arrays create a fresh stream.
    ``timeperiod`` defaults to 20 and controls the trailing mean. Rust owns
    the formula ``0.5*ln(H/L)^2 - (2*ln(2)-1)*ln(C/O)^2``, rolling window, and
    NaN warm-up. ``compute`` returns one aligned float array and ``value`` the
    latest scalar. Lifecycle mutators return ``self``. The independent oracle
    is the Garman-Klass definition used by pandas-ta-classic.
    """

    def __init__(
        self,
        _open: Any,
        high: Any,
        low: Any,
        close: Any,
        timeperiod: int = 20,
    ) -> None:
        """Initialize and process aligned OHLC histories.

        Parameters
        ----------
        _open, high, low, close : object
            Required aligned OHLC histories; empty arrays create a fresh state.
        timeperiod : int, default 20
            Positive trailing window length in bars.
        """
        self._state = _Native(int(timeperiod))
        self.extend(_open, high, low, close)

    def append(
        self, _open: float, high: float, low: float, close: float
    ) -> "GarmanKlass":
        """Append one OHLC bar in open/high/low/close order."""
        self._state.append(float(_open), float(high), float(low), float(close))
        return self

    def extend(
        self, _open: Any, high: Any, low: Any, close: Any
    ) -> "GarmanKlass":
        """Append aligned OHLC histories without mutating on mismatch."""
        arrays = tuple(as_float64_series(value) for value in (_open, high, low, close))
        if len({len(array) for array in arrays}) != 1:
            raise ValueError("_open, high, low, and close must have equal lengths")
        self._state.extend(*arrays)
        return self

    def compute(self) -> np.ndarray:
        """Return the aligned rolling volatility history."""
        return self._state.compute()

    @property
    def value(self) -> float | None:
        """Return the latest volatility, or ``None`` during warm-up."""
        return self._state.value

    def reset(self) -> "GarmanKlass":
        """Restore fresh native state and return this adapter."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of OHLC bars processed by Rust."""
        return adapter_length(self)


__all__ = ["GarmanKlass"]

File: crates/taflow-core/src/indicators/garman_klass.rs
use crate::error::TaResult;
use crate::stream::operator_states::*;

pub struct GarmanKlass {
    mean: RollingMean,
    value: Option<f64>,
}

impl GarmanKlass {
    /// Computes or updates `new` through the native Rust kernel.
    ///
    /// Parameters are the typed series and configuration values in the signature.
    ///
    /// Returns the computed value, aligned history, or a validation error.
    pub fn new(timeperiod: usize) -> TaResult<Self> {
        Ok(Self {
            mean: RollingMean::new(timeperiod)?,
            value: None,
        })
    }

    /// Computes or updates `append` through the native Rust kernel.
    ///
    /// Parameters are the typed series and configuration values in the signature.
    ///
    /// Returns the computed value, aligned history, or a validation error.
    pub fn append(&mut self, open: f64, high: f64, low: f64, close: f64) -> Option<f64> {
        let term = if high > 0.0 && low > 0.0 && open > 0.0 && close > 0.0 {
            0.5 * (high / low).ln().powi(2)
                - (2.0 * 2.0f64.ln() - 1.0) * (close / open).ln().powi(2)
        } else {
            0.0
        };
        self.value = self
            .mean
            .append(term)
            .map(|mean| mean.sqrt() * 252.0_f64.sqrt() * 100.0);
        self.value
    }

    /// Computes or updates `value` through the native Rust kernel.
    ///
    /// Parameters are the typed series and configuration values in the signature.
    ///
    /// Returns the computed value, aligned history, or a validation error.
    pub fn value(&self) -> Option<f64> {
        self.value
    }

    /// Reset the persistent state and clear the latest value.
    pub fn reset(&mut self) {
        self.mean.reset();
        self.value = None;
    }
}
